# Math 548, Midterm. Problem 2. LS

**Problem 2**

**Solve the following systems of linear equations with the Jacobi
iteration method using the initial guess as [ 0, 0, 0 ].**

*   In each case, will the Jacobi iteration converge to a solution? Give the
justification for your answer?
*   If yes, find the solutions.

ii.  

```
     [ 8  3  2][x1]   [ 13]
    
     [ 1  5  1][x2] = [ 7 ]
    
     [ 2  1  6][x3]   [ 9 ]
```



# Finding eigenstuff of a matrix
[Source:](https://pythonnumericalmethods.berkeley.edu/notebooks/chapter15.04-Eigenvalues-and-Eigenvectors-in-Python.html)

In [ ]:
import numpy as np
from numpy.linalg import eig
a = np.array([[0, 2],
              [2, 3]])
w,v=eig(a)
print('E-value:', w)
print('E-vector', v)

E-value: [-1.  4.]
E-vector [[-0.89442719 -0.4472136 ]
 [ 0.4472136  -0.89442719]]


# Solving Ax=b matrix equation
[Source:](https://stackoverflow.com/questions/22163113/matrix-multiplication-solve-ax-b-solve-for-x)

#Problem 2. (ii)

**To check with the actual(real) solution, here, I'm giving the real solution as well:**

In [ ]:
# To solve Ax=b, We start by constructing the arrays for A and b.

A = np.array([[ 8,  3,  2],
              [ 1,  5,  1],
              [ 2,  1,  6]])
b = np.transpose(np.array([ 13,  7,  9]))
# To solve the system we do

x = np.linalg.solve(A,b)
print("Real solution: ")
print(x)

Real solution: 
[1. 1. 1.]


**Using Jacobi Iteration:**

In [ ]:
A = np.array([[ 8,  3,  2],
              [ 1,  5,  1],
              [ 2,  1,  6]])

L = np.array([[ 0,  0,  0],
              [ 1,  0,  0],
              [ 1,  1,  0]])

D = np.array([[ 8,  0,  0],
              [ 0,  5,  0],
              [ 0,  0,  6]])

U = np.array([[ 0,  3,  2],
              [ 0,  0,  1],
              [ 0,  0,  0]])

D_inverse =  np. linalg. inv(D)
b = np.transpose(np.array([ 13,  7,  9]))
D_inverse_b = np.dot(D_inverse, b)

BJ = np.dot(-D_inverse, L+U)
w,v=eig(BJ)

print('BJ evalues:', w)

x0 = np.transpose(np.array([ 0,  0,  0]))
x = list();
x.append(x0)

for i in range(5):
  x1 = np.dot(BJ, x0)+ D_inverse_b
  x0 = x1
  x.append(x1)
Aproximations = np.array(x)
print(Aproximations)
# print((2.083**2 + 2.312**2)**(1/2))

BJ evalues: [-0.44378452  0.26976158  0.17402294]
[[0.         0.         0.        ]
 [1.625      1.4        1.5       ]
 [0.725      0.775      0.99583333]
 [1.08541667 1.05583333 1.25      ]
 [0.9165625  0.93291667 1.143125  ]
 [0.989375   0.9880625  1.19175347]]


*Also pay attention that the spectral radius of matrix BJ,  P(BJ) = 0.44378452 < 1, Jacobi Iteration does converge.*

In [ ]:
# !sudo apt-get install texlive-xetex texlive-fonts-recommended texlive-plain-generic

In [ ]:
# !jupyter nbconvert --to pdf /content/Math548_hw6_Lazizbek.ipynb